Handle Zulip internal links

sirpengi opened the pull request:
Adds support to recognize internal UI links, such as links to other streams or topics.

sirpengi commented:
The `NarrowOperator` enum in `lib/model/internal_link.dart` should be a private class (It should be `_NarrowOperator`) but then the `JsonEnum` annotations would create a `const _$_NarrowOperatorEnumMap` that the analyzer complains about:
```
info: The constant name '_$_NarrowOperatorEnumMap' isn't a lowerCamelCase identifier. (constant_identifier_names at [zulip] lib/model/internal_link.g.dart:11)
```

gnprice commented:
Thanks @sirpengi!

Since I was on vacation the past few days, and then preoccupied today with that Sentry-induced crash bug in zulip-mobile that we fixed with v27.213, here's just some initial comments from a quick skim, for things that will help me review more efficiently when I sit down to a full review.

* Some of this code is transcribed from zulip-mobile. Please mention the specifics of that in the commit messages — which files within zulip-mobile, and which code in the commit (like which functions, or "the test cases", or whatever the case may be).
* I seem to recall there was a test case in zulip-mobile that we discovered in our call last week didn't actually test what it appeared to, and that functionality has actually never worked. (IIRC it was some form of link that was already historical in 2017, when the people who were working on zulip-mobile at that point wrote that test and the corresponding code.) What test case was that?

  Apart from that one, do these tests include all the test cases from zulip-mobile for parsing internal links?

sirpengi commented:
> Thanks @sirpengi!
> 
> Since I was on vacation the past few days, and then preoccupied today with that Sentry-induced crash bug in zulip-mobile that we fixed with v27.213, here's just some initial comments from a quick skim, for things that will help me review more efficiently when I sit down to a full review.
> 
>     * Some of this code is transcribed from zulip-mobile. Please mention the specifics of that in the commit messages — which files within zulip-mobile, and which code in the commit (like which functions, or "the test cases", or whatever the case may be).
> 

Completed, and ready for you to review!

>     * I seem to recall there was a test case in zulip-mobile that we discovered in our call last week didn't actually test what it appeared to, and that functionality has actually never worked. (IIRC it was some form of link that was already historical in 2017, when the people who were working on zulip-mobile at that point wrote that test and the corresponding code.) What test case was that?

This was a DM narrow with an email address (e.g. `/#narrow/dm/a.40b.2Ecom.2Ec.2Ed.2Ecom`) that instead created a narrow object with `NaN`s as operands. The change was also noted in my commit message.

>       Apart from that one, do these tests include all the test cases from zulip-mobile for parsing internal links?

Yes, all the test cases from mobile are also here, except for yet unimplemented narrows (`is/starred`, `is/mentioned`, etc), and some test strings were "translated" (mobile uses "jest" for some stream names and I switched it to "check" in this case).

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,251 @@
+    zulipMergeBase: eg.recentZulipVersion,
+  );
+  final initialSnapshot = eg.initialSnapshot(streams: streams);
+  await testBinding.globalStore.add(account, initialSnapshot);
+  return await testBinding.globalStore.perAccount(account.id);
```
We shouldn't need the global binding here; we're just passing stores around explicitly. See other test files in `test/model/` for examples.

gnprice reviewed (commented):
Thanks!

Here's a full review up through a first tranche: the three prep commits, and then the "parse internal links" commit through the lib code and part of the tests.

Left for a future round are the second half of those tests (the "Get Narrow" group), and the final "handle internal links" commit.

sirpengi commented on lib/model/internal_link.dart:
```diff
@@ -0,0 +1,253 @@
+/// `realm`.
+Narrow? parseInternalLink(Uri url, PerAccountStore store) {
+  if (!url.hasFragment) return null;
+  if (!url.hasScheme || url.host.isEmpty) return null;
```
The `hasScheme` and `host` check were because calling `url.origin` will raise a `StateError` exception if they were not set, although a review of the dart code now shows there are other ways it can fail (see https://github.com/dart-lang/sdk/blob/b8fb4a537c315ae6abe6b8dd5932b1b9adecfbad/sdk/lib/core/uri.dart#L2824) so perhaps I should capture the exception instead. I'll be refactoring this based on some other comments below.

sirpengi commented on lib/model/internal_link.dart:
```diff
@@ -0,0 +1,253 @@
+/// The passed `url` must appear to be a link to a Zulip narrow on the given
+/// `realm`.
+Narrow? parseInternalLink(Uri url, PerAccountStore store) {
+  if (!url.hasFragment) return null;
+  if (!url.hasScheme || url.host.isEmpty) return null;
+  if (!url.hasEmptyPath && (url.path != '/')) return null;
+
+  if (url.origin != store.account.realmUrl.origin) return null;
```
My thinking of putting `hasFragment` up top was for optimization, since all internal links are defined in the fragment, it's a quick check if it exists or not, rather than a string comparision of the origin.

As for query, the mobile app rejects any urls with a query set, but the web UI ignores it. I'll reorder for the general-to-specific (I agree about the added clarity of that) and will follow the mobile behavior on rejecting queries.

sirpengi commented on lib/model/internal_link.dart:
```diff
@@ -0,0 +1,253 @@
+}
+
+@JsonEnum(fieldRename: FieldRename.kebab, alwaysCreate: true)
+enum NarrowOperator {
```
Adding the lint rule worked, and I prefer that situation where it can remain private.

sirpengi commented:
@gnprice ready for review again. I've refactored the test suite to match the zulip-mobile suite in order and made sure every test case was included (and the ones not supported are included but commented out with the appropriate TODO attached if available)

gnprice commented on lib/model/narrow.dart:
```diff
@@ -168,6 +168,13 @@ class DmNarrow extends Narrow implements SendableNarrow {
     );
   }
 
+  factory DmNarrow.withUsers(List<int> userIds, {required int selfUserId}) {
```
nit in commit message:

    Add DmNarrow.withUsers

Include a subject-area prefix: here `narrow:` would be good.

For discussion, see in the Zulip-wide Git style guide: https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-summary-part-1
The examples there are all for the server and web app, but the principle is the same.

gnprice commented on lib/model/internal_link.dart:
```diff
@@ -0,0 +1,282 @@
+
+import 'package:flutter/cupertino.dart';
```
nit: This import is broader than needed. In particular it's a widget library, and this file shouldn't need to know anything about widgets.

Android Studio has a habit of inserting this one when what you actually want can be found in a more base-layer library like `flutter/foundation` or `flutter/widgets`. (Perhaps because "cupertino" comes first alphabetically.)

gnprice commented on lib/model/internal_link.dart:
```diff
@@ -0,0 +1,282 @@
+
+/// Check if `url` is an internal link on the given `realmUrl`.
+bool _isInternalLink(Uri url, Uri realmUrl) {
+  if (url.hasScheme || url.hasAuthority || url.hasPort) {
+    // input that do not include scheme, host, and port
+    // are treated as on realm
```
It really worries me to be adding ad-hoc logic like this in the context of interpreting a URL. That's the sort of thing that creates security vulnerabilities: one system thinks the URL means one thing and refers to something trusted, but then passes the URL to a different system which interprets it as pointing to something else that in fact should not be trusted.

So any time we think there's a need for this sort of logic, we need to be very explicit about thinking through what we're actually trying to do, and why the logic we're writing will correctly do the thing we want and not have edge cases where it goes wrong.

In this instance, we're porting over code from zulip-mobile, and I don't think there's corresponding logic there. So I suspect we ultimately don't need this.

If I take this condition out, I see there are a couple of tests that fail, which I assume is related to why you added it. (And https://github.com/zulip/zulip-flutter/pull/305#discussion_r1334363539 sounds related too.) I recommend starting a chat thread in `#mobile-team` with your observations and debugging of what goes wrong without this condition; that'll be a better medium than GitHub for a debugging or "what's the right way to solve this problem" conversation.

sirpengi replied:
Through our discussion we settled on running all input through `realmUrl.resolve()` so that they have an origin to report later. This matches the mobile behavior of running input through `new URL()` with a base.

gnprice replied:
Linking that chat discussion for cross-reference: https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/URLs.20are.20tricky.20to.20parse/near/1647599

gnprice commented on lib/model/internal_link.dart:
```diff
@@ -0,0 +1,253 @@
+
+  if (url.origin != store.account.realmUrl.origin) return null;
+
+  final (category, segments) = _getCategoryAndSegmentsFromFragment(url.fragment);
```
OK, works for me.

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,512 @@
+      final String because = testCase.$2;
+      final Uri url = testCase.$3;
+      final Uri realmUrl = testCase.$4;
+      test(because, () async {
```
The zulip-mobile version has a bit more in the test description:
```js
    test(`${expected ? 'accept' : 'reject'} ${description}: ${url.toString()}`, () => {
```

I think the accept/reject part, especially, is helpful when you're looking at a list of these that failed, or passed, and trying to understand what they're saying.

sirpengi commented on lib/model/internal_link.dart:
```diff
@@ -0,0 +1,282 @@
+      case _NarrowOperator.subject:
+        if (topicElement != null) return null;
+        final String topic;
+        try {
+          topic = decodeHashComponent(operand);
+        } on ArgumentError {
+          return null;
+        } on FormatException {
+          return null;
+        }
```
This method relies on `Uri.decodeComponent` which can be easily made to throw when you pass in the wrong percentage encoding values:
```
  final cases = [
    "%1",
    "%FF",
  ];
  for (final input in cases) {
    try {
      Uri.decodeComponent(input);
    } catch (e) {
      print("$input: ${e.toString()}");
    }
  }
```

gnprice replied:
Cool, thanks for the examples. I guess that's the same behavior as JS's (or the Web platform's?) `decodeURIComponent` — from my browser console:
```
> decodeURIComponent('%1');
VM99:1 Uncaught URIError: URI malformed
    at decodeURIComponent (<anonymous>)
    at <anonymous>:1:1
> decodeURIComponent('%FF');
VM106:1 Uncaught URIError: URI malformed
    at decodeURIComponent (<anonymous>)
    at <anonymous>:1:1
```

Kind of messy that `Uri.decodeComponent` throws two different exception types there, but so it goes.

Anyway, should be encapsulated inside our `decodeHashComponent` helper rather than having each caller have to deal with it.

gnprice replied:
Oh and to complete your examples — here's the output:
```
%1: Invalid argument(s): Truncated URI
%FF: FormatException: Invalid UTF-8 byte (at offset 0)
```

sirpengi commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,512 @@
+  List<User>? users,
+}) async {
+  selfUser ??= eg.selfUser;
+  final account = Account(
+    id: 1001,
+    realmUrl: realmUrl,
```
For these tests we need to create various accounts, sometimes with a custom `realmUrl`. I found an `Account.copyWith` that will make this clear what we're replacing outside the baseline test data.

sirpengi commented:
@gnprice this is ready for review again!

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,489 @@
+        Uri.parse('https://example.com:444/#narrow/stream/check'),
+        Uri.parse('https://example.com:444/')),
+
+      // In the following examples unicode domains will
+      // pass but punycode variants of unicode domains are not
+      // treated as matching because Dart's [Uri] currently
```
This description "In the following … matching" isn't quite right. Punycode is involved only in the first of the three failing examples; and it's not a variant, but an encoding.

I think the description can actually just be left out, though — the examples speak for themselves about what doesn't work. What this comment can add is the next few lines, which point to why they don't work.

(cf https://github.com/zulip/zulip-flutter/pull/305#discussion_r1336378571 )

gnprice commented on lib/model/internal_link.dart:
```diff
@@ -0,0 +1,285 @@
+
+  for (var i = 0; i < segments.length; i += 2) {
+    final (operator, negated) = _parseOperator(segments[i]);
+    // TODO(#252): negated operators are not currently supported
+    // These should eventually return a SearchNarrow
+    if (negated) return null;
```
```suggestion
    // TODO(#252): negated operators are not currently supported;
    //   they should eventually return a SearchNarrow
    if (negated) return null;
```

That is, I _think_ the intent here is that the second comment line here is elaborating on the TODO, and "these" refers to negated operators; this would be a way of making that explicit. In particular the indentation groups it into the TODO.

If "these" is intended to mean something else then I guess it'd be good to make that explicit instead :-)

gnprice replied:
Hmm and then reading a bit more of the code around this: there are several other spots in this function where a similar comment applies, right?

I think the biggest one is probably the `_NarrowOperator.unknown` case — that's what covers plain search words, and `sender:`, and miscellaneous other search features. But also having multiple stream, topic, or DM elements; or having a DM element and a stream or topic element; and maybe some of the other `return null` cases too.

So the cleanest way to express this point may be to put a comment up somewhere like the top of the function. For example… oh, in fact we already have a discussion of this up at `parseInternalLink`. So can just add the TODO (with the handy issue-number reference) there, like:
```diff
 /// This can also return null for some valid narrow links that our Narrow
 /// type *could* accurately represent. We should try to understand these
 /// better, but some kinds will be rare, even unheard-of:
 ///   #narrow/stream/1-announce/stream/1-announce (duplicated operator)
+// TODO(#252): handle all valid narrow links, returning a search narrow
 Narrow? parseInternalLink(Uri referenceUrl, PerAccountStore store) {
```

I think that covers it. If we set out to implement that TODO, we'll naturally go into `_interpretNarrowSegments` and find all the places where it's returning null instead of a narrow, and identify whether each one corresponds to a valid narrow link and so should start returning a search narrow instead.

gnprice commented on lib/model/internal_link.dart:
```diff
@@ -0,0 +1,285 @@
+@visibleForTesting
+String? decodeHashComponent(String str) {
+  try {
+    return Uri.decodeComponent(str.replaceAll('.', '%'));
+  } on ArgumentError {
+    // as with '%1': Invalid argument(s): Truncated URI
+    return null;
+  } on FormatException {
+    // as with '%FF': FormatException: Invalid UTF-8 byte (at offset 0)
```
Cool, these comments are helpful — basically documenting in the code the things you learned at https://github.com/zulip/zulip-flutter/pull/305#discussion_r1340487666 , which aren't in upstream's documentation.

gnprice commented on lib/model/internal_link.dart:
```diff
@@ -0,0 +1,285 @@
+/// Create a new `Uri` object in relation to a given realmUrl.
+///
+/// Returns `null` if `urlString` could not be parsed as a `Uri`.
+Uri? tryResolveToRealmUrl(String urlString, Uri realmUrl) {
```
This function's name says it tries to resolve the argument *to* a realm URL. That'd mean the output, if any, would be a realm URL.

But that's not what it does; it takes the realm URL as an input, and the output is something else — could be the realm URL again if `urlString` is e.g. "/", or could be some other URL on the realm if `urlString` is some other relative-URL string like "#narrow/", or could be a URL somewhere else entirely if `urlString` is an absolute-URL string like "https://elsewhere.example/".

Could say "on" instead of "to"; that'd be enough to make it accurate.

gnprice commented on lib/model/internal_link.dart:
```diff
@@ -0,0 +1,285 @@
+  return store.account.realmUrl.replace(fragment: fragment.toString());
+}
+
+/// Create a new `Uri` object in relation to a given realmUrl.
+///
+/// Returns `null` if `urlString` could not be parsed as a `Uri`.
+Uri? tryResolveToRealmUrl(String urlString, Uri realmUrl) {
```
As is, this is just called in one place, and it's not mentioned in any docs. Did you intend to refer to it somewhere, like as part of describing the requirements `parseInternalLink` makes on its caller?

sirpengi replied:
Yes, I've added a note in `parseInternalLink` to refer to this

gnprice commented on lib/model/internal_link.dart:
```diff
@@ -0,0 +1,285 @@
+    .replaceAllMapped(_encodeHashComponentRegex, (Match m) => _hashReplacements[m[0]!]!);
+}
+
+/// Decode a dot-encoded string.
+// The Zulip webapp uses this encoding in narrow-links:
+// https://github.com/zulip/zulip/blob/1577662a6/static/js/hash_util.js#L18-L25
+@visibleForTesting
+String? decodeHashComponent(String str) {
```
```suggestion
/// Decode a dot-encoded string; return null if malformed.
// The Zulip webapp uses this encoding in narrow-links:
// https://github.com/zulip/zulip/blob/1577662a6/static/js/hash_util.js#L18-L25
@visibleForTesting
String? decodeHashComponent(String str) {
```

gnprice commented on lib/model/internal_link.dart:
```diff
@@ -0,0 +1,285 @@
+/// type *could* accurately represent. We should try to understand these
+/// better, but some kinds will be rare, even unheard-of:
+///   #narrow/stream/1-announce/stream/1-announce (duplicated operator)
+Narrow? parseInternalLink(Uri referenceUrl, PerAccountStore store) {
+  final url = store.account.realmUrl.resolveUri(referenceUrl);
```
Hmm interesting. Why the additional `resolveUri` call here?

This makes a significant difference in what the meaning of this function is. It definitely needs to be clearly reflected in the function's doc.

But also in the way this function gets called, it seems redundant — we already pass it a URL that's been resolved to an absolute URL.

It looks like there are no tests exercising this behavior, either — this diff passes all the tests:
```diff
 Narrow? parseInternalLink(Uri referenceUrl, PerAccountStore store) {
-  final url = store.account.realmUrl.resolveUri(referenceUrl);
+  final url = referenceUrl;
```
So if we were to have this behavior, we would definitely want to have tests for it.

sirpengi replied:
There was an oversight on my part, it should not be called again. Along with a doc change in https://github.com/zulip/zulip-flutter/pull/305#discussion_r1347181062 this line will be removed.

gnprice reviewed (commented):
Thanks @sirpengi for the revision! I've gone through all the review threads from the last revision, and all the code except the last commit and `internal_link_test.dart`. This is definitely getting closer; just a few comments, below and above (https://github.com/zulip/zulip-flutter/pull/305#pullrequestreview-1657966380), and most of them small.

For a subsequent review I'll read through the rest of the tests again, and the final commit "content: Handle internal links".

sirpengi commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,489 @@
+    List<ZulipStream>? streams,
+    PerAccountStore? store,
+    List<User>? users,
+  }) async {
```
I was able to remove `async` from the whole test

sirpengi commented:
@gnprice ready again!

gnprice commented on lib/model/internal_link.dart:
```diff
@@ -0,0 +1,285 @@
+/// Create a new `Uri` object in relation to a given realmUrl.
+///
+/// Returns `null` if `urlString` could not be parsed as a `Uri`.
+Uri? tryResolveOnRealmUrl(String urlString, Uri realmUrl) {
```
This commit message says:

    internal_link: Add tryParseToRealmUrl

Should match the name that's in the code.

gnprice replied:
bump

(I guess it's now `internal_link: Add tryParseOnRealmUrl` but that still differs from the code :-) )

gnprice reviewed (commented):
Thanks @sirpengi! Reading these revisions, they all look good except one nit below.

Next I'll do the remaining reading mentioned above at https://github.com/zulip/zulip-flutter/pull/305#pullrequestreview-1658052878 :
> For a subsequent review I'll read through the rest of the tests again, and the final commit "content: Handle internal links".

though I might not get to that today.

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,486 @@
+      eg.stream(streamId: 1, name: "general"),
+    ];
+
+    test('basic', () {
+      final store = setupStore(realmUrl: realmUrl, streams: streams);
+      final testCases = [
+        ('#narrow/stream/1-general', const StreamNarrow(1)),
+      ];
+      checkExpectedNarrows(testCases, store: store);
+    });
+
+    test('if stream not found, use stream ID anyway', () {
+      final store = setupStore(realmUrl: realmUrl, streams: streams);
+      final testCases = [
+        ('#narrow/stream/123-topic', const StreamNarrow(123)),
+      ];
+      checkExpectedNarrows(testCases, store: store);
+    });
```
Compare these lines to the corresponding lines in the zulip-mobile tests:

```js
    test('basic', () => {
      expectStream(`${streamGeneral.stream_id}-general`, [streamGeneral], streamGeneral);
    });

    test('if stream not found, use stream ID anyway', () => {
      expectStream(`${streamGeneral.stream_id}-general`, [], streamGeneral);
    });
```

In the latter, the two lines with the details of the tests are separated by 1 line of text and 2 nearly-blank lines. In this version, they're separated by 7 lines, of which 4 have text on them. That makes it a lot noisier for the eye to compare the two and see what the contrasts are. Similarly for the next few cases below.

For this sort of thing, it can really help a lot to make a small local helper function. That's what `expectStream` in the corresponding zulip-mobile code is doing.

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,486 @@
+      checkExpectedNarrows(testCases, store: store);
+    });
+
+    test('on malformed stream link: reject', () {
+      final store = setupStore(realmUrl: realmUrl, streams: streams);
+      final testCases = [
+        ('#narrow/stream/-1', null),
+        ('#narrow/stream/1nonsense-general', null),
+        ('#narrow/stream/-general', null),
```
(Here it's OK that these are all in one `test` call, because that's not a regression from the zulip-mobile tests.)

gnprice reviewed (commented):
Thanks again @sirpengi! I've now read through much of the remaining tests; comments below.

I haven't yet read through all of the tests closely, because this seems like enough review comments for one round. As you act on these comments, though, please do look at the rest of the file to see where similar patterns appear, and apply similar revisions there where appropriate.

I did skim through the rest of these tests, and I see they're mostly in a similar structure to the tests I read in this round — so I think once you apply these comments through the whole file, it should all be quite close to merge.

(NB there's also that one small comment just above at https://github.com/zulip/zulip-flutter/pull/305#pullrequestreview-1660968006 , as the first part of this round of review.)

sirpengi commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,486 @@
+      eg.stream(streamId: 123, name: 'topic'),
+    ];
+
+    test('/#narrow/stream/... returns expected StreamNarrow', () {
+      final store = setupStore(realmUrl: realmUrl, streams: streams);
+      final testCases = [
+        ('/#narrow/stream/check', const StreamNarrow(1)),
+        ('/#narrow/stream/stream/', const StreamNarrow(5)),
+        ('/#narrow/stream/topic/', const StreamNarrow(123)),
+      ];
+      checkExpectedNarrows(testCases, store: store);
```
I've found out that it's okay to have multiple layers of groups, so I've refactored the tests that each test case becomes its own `test()` call.

gnprice replied:
Great, thanks.

And yeah, in general, if there's something that was asked for (in a review, or in the issue description, etc.) that doesn't seem to have been possible to do, please be sure to flag it explicitly. That way we can discuss alternatives and tradeoffs — or perhaps someone will know a way it can be done after all.

sirpengi commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,486 @@
+    });
+
+    test('/#narrow/stream/.../topic/... returns expected TopicNarrow', () {
+      final store = setupStore(realmUrl: realmUrl, streams: streams);
```
I've refactored `checkExpectedNarrows` to also take in a `realmUrl` in order to integrate these test cases as well.

sirpengi commented:
@gnprice ready for review again. I've also in other areas tried to compact the tests as much as possible so that it's more legible in spirit with the comments provided.

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,394 @@
+    }
+  }
+
+  group('ParseInternalLink', () {
```
nit: being in camel case this looks like an identifier, not normal English; but we don't have any identifier by this name.

sirpengi replied:
Meant to have been `parseInternalLink`

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,394 @@
+void main() {
+  final realmUrl = Uri.parse('https://example.com/');
+
+  checkExpectedNarrows(List<(String, Narrow?)> testCases, {
```
nit: since this now contains a `test` call, its name should say "test" rather than "check" — that convention helps when reading callers, for seeing the nesting of `test` calls

(I see the corresponding test file in zulip-mobile doesn't get that right, oops.)

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,394 @@
+      (false, 'triple slash', realmUrl,
+        Uri.parse('https://example.com///#narrow/stream/check')),
+
+      (true, 'with port', Uri.parse('https://example.com:444/'),
+        Uri.parse('https://example.com:444/#narrow/stream/check')),
+
+      // Dart's [Uri] currently lacks IDNA or Punycode support:
+      //   https://github.com/dart-lang/sdk/issues/26284
+      //   https://github.com/dart-lang/sdk/issues/29420
+
+      // (true, 'same domain, punycoded host', Uri.parse('https://example.भारत/'),
+      //   Uri.parse('https://example.xn--h2brj9c/#narrow/stream/check')), // FAILS
+
+      (true, 'punycodable host', Uri.parse('https://example.भारत/'),
+        Uri.parse('https://example.भारत/#narrow/stream/check')),
+
+      // (true, 'same domain, IDNA-mappable', Uri.parse('https://example.com'),
+      //   Uri.parse('https://ℯⅩªm🄿ₗℰ.ℭᴼⓂ/#narrow/stream/check')), // FAILS
+
+      (true, 'ipv4 address', Uri.parse('http://192.168.0.1/'),
+        Uri.parse('http://192.168.0.1/#narrow/stream/check')),
+
+      // (true, 'same IPv4 address, IDNA-mappable', Uri.parse('http://192.168.0.1/')),
+      //   Uri.parse('http://１𝟗𝟚。①⁶🯸．₀｡𝟭/#narrow/stream/check')), // FAILS
```
I think these are all more readable the way they were (which is how we have them in the zulip-mobile tests): the action is all happening in the authorities (the hosts + ports), particularly in comparing the realm URL's authority to the input URL's authority. So having those two aligned with each other is best for that.

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,394 @@
+    });
+
+    group('unexpected link shapes are rejected', () {
+      final testCases = [
+        ('/#narrow/stream/name/topic/',           null),
+        ('/#narrow/stream/name/unknown/operand/', null),
```
```suggestion
      final testCases = [
        ('/#narrow/stream/name/topic/',           null), // missing operand
        ('/#narrow/stream/name/unknown/operand/', null), // unknown operator
```

(like the comments in the zulip-mobile version)

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,394 @@
+    store ??= setupStore(realmUrl: realmUrl, streams: streams, users: users);
+    for (final testCase in testCases) {
+      final String urlString = testCase.$1;
+      final Uri url = realmUrl.resolve(urlString);
+      final Narrow? expected = testCase.$2;
+      test(urlString, () {
+        check(parseInternalLink(url, store!)).equals(expected);
```
This test helper is violating the contract prescribed in the function's docs:
```dart
/// A [Narrow] from a given URL, on `store`'s realm.
///
/// `url` must already be passed through [tryResolveOnRealmUrl].
```

We should not do that — it undermines the value of the test.

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,394 @@
+      checkExpectedNarrows(testCases, streams: streams);
+    });
+
+    group('on old stream link, without realm info', () {
+      final stream = eg.stream(name: 'example');
+      final testCases = [
+        ('/#narrow/stream/${stream.name}/', StreamNarrow(stream.streamId)),
+        ('#narrow/stream/${stream.name}/',  StreamNarrow(stream.streamId)),
```
The description says "without realm info", but these don't have any less info about the realm than the tests above do.

In the corresponding zulip-mobile tests, the difference is that the other group has absolute URL strings like `https://example.com/#narrow/stream/jest`, while this has relative URL strings. That's already a fairly unnecessary distinction because the test helper there ultimately does `new URL(url, 'https://example.com')` before passing the URL to the code under test; the distinction dates from when the underlying code under test actually took a URL string instead of a `URL` object.

I think these two cases don't add anything not covered in the group above, so we can just drop this one.

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,394 @@
+
+  group('ParseInternalLink', () {
+    group('topic link parsing', () {
+      final stream = eg.stream(name: "example");
+
+      group('basic', () {
+        String mkUrl(operand) {
+          return '#narrow/stream/${stream.streamId}-general/topic/$operand';
```
This is a bit incongruous: the stream's name is "example" but the URL says "general". Typically the URL would accurately reflect the stream name. Our tests should generally use representative data except when the test is specifically about some way of being non-representative, so these should agree.

(And other tests above cover the possibility that the URL doesn't reflect the stream name.)

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,394 @@
+        checkExpectedNarrows(testCases, streams: [stream]);
+      });
+
+      group('on old topic link, without realm info', () {
+        final testCases = [
+          ('/#narrow/stream/${stream.name}/topic/topic', TopicNarrow(stream.streamId, 'topic')),
+          ('#narrow/stream/${stream.name}/topic/topic',  TopicNarrow(stream.streamId, 'topic')),
```
Similar to the stream case above, this description doesn't really fit, and this pair of cases can just be dropped.

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,394 @@
+          return '#narrow/stream/${stream.name}/topic/$operand';
+        }
+        final testCases = [
+          (mkUrl('(no.20topic)'), TopicNarrow(stream.streamId, '(no topic)')),
+          (mkUrl('google.2Ecom'), TopicNarrow(stream.streamId, 'google.com')),
+          (mkUrl('google.com'),   null),
+          (mkUrl('topic.20name'), TopicNarrow(stream.streamId, 'topic name')),
```
Neat, I appreciate the compactness here.

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,394 @@
+        checkExpectedDmNarrow('#narrow/pm-with/1,2-group');
+      });
+
+      group('on group PM link including self', () {
+        checkExpectedDmNarrow('#narrow/dm/1,2,${eg.selfUser.userId}-group');
```
The zulip-mobile version has a comment that's important context for understanding this test:
```suggestion
      group('on group PM link including self', () {
        // The webapp doesn't generate these, but best to handle them anyway.
        checkExpectedDmNarrow('#narrow/dm/1,2,${eg.selfUser.userId}-group');
```

gnprice reviewed (commented):
Thanks for the revision! As I hoped after the last review, this is now getting quite close to merge. I've read through the whole test file, and just have the handful of comments below and above (https://github.com/zulip/zulip-flutter/pull/305#pullrequestreview-1665329750), all small.

Because this PR thread has gotten long, let's break the last commit:
f6ab5e7e1 content: Handle internal links
out to a separate followup PR. That'll be a much shorter thread, I think, but probably not totally trivial, so best to start it fresh.

sirpengi commented:
@gnprice ready to go again! I have split off the extra integration commit and will open another pr after this

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,387 @@
+    });
+
+    group('DM link parsing', () {
+      testExpectedDmNarrow(String testCase) {
+        final expectedNarrow = DmNarrow.withUsers([1, 2],
+          selfUserId: eg.selfUser.userId);
+        testExpectedNarrows([(testCase, expectedNarrow)], users: [
```
Hmm, this looks wrong — it's a "test" function inside another "test" function.

… Oh, aha. This is a function *definition*, not a function call, of `testExpectedDmNarrow`. So that nesting is fine, then.

Let's give this an explicit return type (of `void`). That's good for function definitions in general because it makes the expectations explicit; and then I guess another benefit I hadn't previously appreciated is that it makes them more visually distinct from function calls.

sirpengi replied:
Applied to the same to `testExpectedStreamNarrow` above this

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,387 @@
+        'https://example.भारत/#narrow/stream/check',
+        Uri.parse('https://example.भारत/')),
+
+      // (true, 'same domain, IDNA-mappable', Uri.parse('https://example.com'),
+      //   'https://ℯⅩªm🄿ₗℰ.ℭᴼⓂ/#narrow/stream/check'), // FAILS
```
nit: this one is flipped around compared to the others

gnprice reviewed (commented):
Thanks @sirpengi for the revision! All looks great now except three small comments below — after those are fixed, this should be all ready to merge.

Please go ahead and send the followup PR too, as the branch that includes these changes and then that one further commit. Just mention in the PR description that it's on top of this PR.

sirpengi commented:
@gnprice good to go again!

sirpengi commented:
For reference, the trailing commit is in #318

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,388 @@
+void main() {
+  final realmUrl = Uri.parse('https://example.com/');
+
+  testExpectedNarrows(List<(String, Narrow?)> testCases, {
```
Same thing as at https://github.com/zulip/zulip-flutter/pull/305#discussion_r1355788136 .

gnprice commented on test/model/internal_link_test.dart:
```diff
@@ -0,0 +1,388 @@
+        'https://example.भारत/#narrow/stream/check',
+        Uri.parse('https://example.भारत/')),
+
+      // (true, 'same domain, IDNA-mappable'
+      //   'https://ℯⅩªm🄿ₗℰ.ℭᴼⓂ/#narrow/stream/check'),
+      //   Uri.parse('https://example.com'), // FAILS
```
This version doesn't parse if you uncomment it.

Looking at the neighboring commented-out failures, they don't either (but more subtly) — there's a stray close-paren.

gnprice reviewed (commented):
Thanks! Couple of spots remaining from the last review; see below.

And thanks for sending #318 — I'll take a look at that next.

sirpengi commented:
@gnprice ready again

gnprice commented:
Thanks again @sirpengi for all the revisions!

This now all looks good — merging.
